text/0000-pi-types.md

```diff
@@ -0,0 +1,301 @@
+
+Structural equality plays a key role in type checking of dependent types.
+
+Structural equality, in this case, is defined as an equivalence relation, which
+allows substitution without changing semantics.
+
+Any constant parameter must have the `structural_match` property as defined in
+[RFC #1445](https://github.com/rust-lang/rfcs/pull/1445). This property, added
+through the `#[structural_match]` attribute, essentially states that the `Eq`
+implementation is structural.
+
+Without this form of equality, substitution wouldn't be possible, and thus
+typechecking an arbitrarily value-depending type constructor would not be
+possible.
+
+## Parsing
+
+Introducing expr subgrammar in type position isn't possible without setting
+some restrictions to the possible expressions.
+
+In this case, we put restrictions to arithmetic and bitwise operations (`+-/!^`
+etc.) and function calls (`myconstfn(a, b, c, ...)` with `a, b, c, ...` being
+`const-expr` fragments).
+
+Additionally, we allow parenthesizes, which can contain any general const-expr
+fragment.
+
+# How we teach this
+
+This RFC aims to keep a "symmetric" syntax to the current construct, giving an
+intuitive behavior, however there are multiple things that are worth explaining
+and/or clearing up:
+
+**What are dependent types?**
+
+Dependent types are types, which _depend_ on values, instead of types. For
+example, [T; 3], is dependent since it depends on the value, `3`, for
+constructing the type. Dependent types, in a sense, are similar to normal
+generics, where types can depend on other types (e.g. `Vec<T>`), whereas
+dependent types depend on values.
+
+**How does this differ from other languages' implementations of dependent types?**
+
+Various other languages have dependent type systems. Strictly speaking, all
+that is required for a dependent type system is value-to-type constructors,
+although some languages (coq, agda, etc.) goes a step further and remove the
+boundary between value and type. Unfortunately, as cool as it sounds, it has
+some severe disadvantages: most importantly, the type checking becomes
+undecidable. Often you would need some form of theorem prover to type check
+the program, and those have their limitations too.
+
+**What are `const fn` and how is it linked to this RFC?**
+
+`const fn` is a function, which can be evaluated at compile time. While it
+is currently rather limited, in the future it will be extended (see
+[Miri](https://github.com/solson/miri)). You can use constexprs to take one
+type-level value, and non-trivially calculate a new one.
+
+**What are the usecases?**
+
+There are many usecases for this. The most prominent one, perhaps, is
+abstracting over generically sized arrays. Dependent types allows one to lift
+the length of the array up to the type-level, effectively allowing one to
+parameterize over them.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If we want to have type-level Turing completeness, the halting problem is
```

Comment:
What I meant was that we should avoid making these errors more likely to come up if possible. I can imagine new users getting such an error and just thinking that the compiler is really slow or poorly engineered, rather than realizing that they are actually asking the compiler to do something impossible.